assets/sum-my-subscriptions/unhappy_flows.md

```diff
@@ -0,0 +1,35 @@
+# Sequence diagrams of the unhappy flows
+
+>## Notes - to be removed
+>What looping do we want in the failure cases:
+>- Should director go back to task agent and ask for a different task breakdown?
+>- Should the router try and look for an agent again?
+>- Should the router continue and try its best to solve the remaining tasks and >let the director know which tasks it couldn't solve?
+
+The below assume no looping for error cases:
+
+## "Router unable to find agent to solve task"
+```mermaid
+sequenceDiagram
+    Router -->> LLM: "Determine which agent is best for this task"
+    LLM -->> Router: "No suitable agent found"
+    Router -->> Supervisor: "No more avaialble agents, task failed"
+    Supervisor -->> Director: "Unable to solve task: 'task_description'"
+    Director -->> frontend: "I'm sorry, I was unable to solve that task"
+```
+
+## "Router choses agent, unable to solve task, no more agents"
+```mermaid
+sequenceDiagram
+    Router -->> LLM: "Determine which agent is best for this task"
+    LLM -->> Router: "Use DatastoreAgent"
+    Router -->> Supervisor: "Call DatastoreAgent with prompt 'get Amazon spending'"
+    Supervisor -->> DatastoreAgent: "Find the answer to the following: 'get Amazon spending'"
+    DatastoreAgent -->> LLM: "Function call: identify methods"
+    LLM -->> DatastoreAgent: "No appropriate method found"
+    DatastoreAgent -->> Supervisor: "I cannot solve this task"
+    Supervisor -->> Router: "Agent failed, try again"
+    Router -->> Supervisor: "No more avaialble agents, task failed"
+    Supervisor -->> Director: "Unable to solve task: 'task_description'"
+    Director -->> frontend: "I'm sorry, I was unable to solve that task"
+```

```

Comment:
A couple of thoughts: 
1. Is there another path where the agent calls the correct function and then the function fails in some way ie. db connection doesn't work? How would this be handled. Would it try again or just say failed?
2. If one task fails do we move onto the next task or does the whole thing fail? Does the flow need to show all tasks?

Reply:
For the first point, in the sequence diagram there is no flow where a correct method is identified and then still fails ie. carrying on to the database layer (not shown in the diagram) and failing there.

In this scenario do we try the same agent again? Do we try and get better arguments for the query? I feel a level of looping that may happen hasn't been included. Not sure if this detail has been confirmed yet though so happy to discuss

Reply:
We're comparing the scenario where we fail a task and fail overall and attempt to do a task n times and fail each time, then eventually failing overall. I think the `loop` notation would probably be appropriate here even if it complicates the graph a little bit but also happy to leave as it is as in the end the behaviour is the same.